Add Impersonate to System.Security.Principal.WindowsIdentity

MarcoRossignoli opened the pull request:
contributes to https://github.com/dotnet/corefx/issues/17164

@danmosemsft for the moment I added Windows test account creator directly on test to avoid too big PR, after merge I could open issue to track effort to move to test lib helper and to amend named pipe tests.
I added a test on different behaviour between `RunImpersonate` and `Impersonate` with regard to `ExecutionContext`

/cc @jkotas @Anipik

jkotas commented on src/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs:
```diff
@@ -208,20 +217,74 @@ public WindowsIdentity(string sUserPrincipalName)
+                    }
+                }
+            }
+            catch
```
You can avoid the `privilege.Revert();` code duplication here by swapping the order of finally and catch. Like:

```
finally
{
    privilege?.Revert();
}
catch
{
    // protect against exception filter-based luring attacks
    throw;
}
```

MarcoRossignoli replied:
I'll wait full review to avoid too much commit.

stephentoub replied:
> You can avoid the privilege.Revert(); code duplication here by swapping the order of finally and catch

That's not valid C#, is it?

jkotas replied:
Right, there would need to be an extra scope:

```csharp
try
{
    try
    {
         do stuff
    }
    finally 
    {
         revert
    }
}
catch
{
     // protect against exception filter-based luring attacks
    throw;
}
```

MarcoRossignoli replied:
I didn't try code above...but maybe the idea was
```cs
...
try
{
   Privilege privilege;
   try
   {
      ...
   }
   finally
   {
     privilege?.Revert();
   }
}
catch
{
   throw;
}
```

jkotas commented:
@danmosemsft Could you please find somebody who understands Windows Security to review this?

MarcoRossignoli commented:
System.DirectoryService is not supported on Nano https://github.com/dotnet/corefx/issues/30916 should I disable tests on it, or with direct pinvoke we should make it works?

GrabYourPitchforks commented:
How does this interact with recent APIs like `UnsafeQueueUserWorkItem`? Since we're flowing the impersonation information through an async local rather than the execution context, does this mean that there's a different behavior in netfx vs. netcoreapp regarding when impersonation is flowed vs. when it's not flowed?

danmosemsft commented:
> System.DirectoryService is not supported on Nano #30916 should I disable tests on it, or with direct pinvoke we should make it works?

Ideally pinvoke to make sure it works. If that is very laborious then I would be happy with disabling the test for Nano with an open issue.

MarcoRossignoli commented:
> How does this interact with recent APIs like UnsafeQueueUserWorkItem? Since we're flowing the impersonation information through an async local rather than the execution context, does this mean that there's a different behavior in netfx vs. netcoreapp regarding when impersonation is flowed vs. when it's not flowed?

@GrabYourPitchforks I'll do some check
However I discussed with @jkotas(https://github.com/dotnet/corefx/issues/17164#issuecomment-440008477) and today there is already some difference, netcoreapp save/restores EC, netfx doesn't do that.
Netfx [`RunImpersonate`](https://referencesource.microsoft.com/#mscorlib/system/security/principal/windowsidentity.cs,614) is a wrapper to `WindowsImpersonationContext`  that [save/restores](https://referencesource.microsoft.com/#mscorlib/system/security/principal/windowsidentity.cs,775) `WindowsIdentity` directly on context for CAS.

MarcoRossignoli commented:
I think there is another difference, `RunImpersonate` uses EC to run delegate. In this way if on delegate there are some asyn/await and we flow context on new thread [`CurrentImpersonatedTokenChanged`](https://github.com/dotnet/corefx/blob/master/src/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs#L750) re-impersonate with correct token new thread(and revert impersonation on old thread). Netfx doesn't guarantee this, if I understand well, or there is some intrinsic mechanism?I mean if we come back on different thread we need to call back `ImpersonateLoggedOnUser`, right?
Again, how I can capture and restore EC like `RunImpersonate`?There isn't an api `ExectutionContext.Restore(ExecutionContext ec)`, here we don't have a delegate.

MarcoRossignoli commented:
@danmosemsft weird...I used `netapi32.dll` that seems supported by nano https://msdn.microsoft.com/en-us/library/mt588480%28v=vs.85%29.aspx?f=255&MSPPError=-2147217396 but I get error `Unable to load DLL 'netapi32.dll' or one of its dependencies: The specified module could not be found. (Exception from HRESULT: 0x8007007E)` or `Windows.10.Nano.Amd64.Open-x64-Debug` leg is different SO?
Question...is there a doc with all available legs explained?

danmosemsft commented:
@MarcoRossignoli there is not a doc describing the various legs. 

>  I get strange fail test on Windows.10.Amd64.ClientRS4.Open-x86-Release but not on other Windows.10.* I would understand the difference.

That is the leg that tests against .NET Framework rather than .NET Core. (We should fix the name. I am not sure where it is set.) The way I know for sure is to look at the [execution logs\](https://mc.dot.net/#/user/MarcoRossignoli/pr~2Fjenkins~2Fdotnet~2Fcorefx~2Fmaster~2F/test~2Ffunctional~2Fcli~2F/a7fae8af8ecdff6b55b1a4955116e8ec2e5b8467/workItem/System.Security.Principal.Windows.Tests/wilogs) for the run and look for `-notrait category=nonnetfxtests` (or just remember the name..)

If you are getting failures specifically on that leg then most likely the tests you wrote are expecting behavior that is different to .NET Framework. You would either want to fix the .NET Core code you wrote, or make the tests skip for .NET Framework. You can run against .NET Framework locally. The [dev guide ](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/developer-guide.md) looks like it needs updating (we are making some changes) but `build -targetgroup netfx` is the way to start, I think.

For Nano - not sure. If you are eager to investigate, I think you can run Nano in a docker image. These are the notes I made last time I did this:
```
1) Install Docker for Windows
2) Make sure to put it in Windows Container mode

3) From an Admin command prompt:
 
a. get the image:
 
	docker pull microsoft/powershell:6.0.2-nanoserver-1709
 
b. start the image in a cmd shell interactively, mounting the local "C:\scratch\mount" to C:\mount inside the container:

	docker run -v c:/scratch/mount:c:/mount -it --rm microsoft/powershell:6.0.2-nanoserver-1709 c:\Windows\system32\cmd.exe

4) when done, call exit
5) when really done: 'docker system prune'
```
However as mentioned I think it is fine to disable the test on Nano for now.

MarcoRossignoli commented:
> That is the leg that tests against .NET Framework rather than .NET Core. (We should fix the name. I am not sure where it is set.) The way I know for sure is to look at the execution logs for the run and look for -notrait category=nonnetfxtests (or just remember the name..)

@danmosemsft thank's a lot for help...some doc on legs could be useful for contributors to became more "autonomous" and waste less time...at least for me :-). Sometimes I run netfx build/tests...but this time I forgot leg/framework association :expressionless:. Thank's for new guide https://github.com/dotnet/coreclr/blob/master/Documentation/project-docs/changing-corelib.md and for details.

> However as mentioned I think it is fine to disable the test on Nano for now.

Ok for the moment I give up on it, great tips on nano on docker also, thank's!I put it in my toolbox.

stephentoub commented on src/Common/src/System/Security/AccessControl/PrivilegeNotHeldException.cs:
```diff
@@ -10,7 +10,12 @@ namespace System.Security.AccessControl
+#else
+    internal
+#endif
+        sealed class PrivilegeNotHeldException : UnauthorizedAccessException, ISerializable
```
This is a little strange.  Do we allow this exception to propagate out of the assembly that this gets built into as internal?  If so, that'll lead to a developer writing code like:
```C#
catch (PrivilegeNotHeldException) { ... }
```
and seeing `PrivilegeNotHeldException` exceptions not being caught by it, because they're not unified and are actually two completely different exception types, albeit with the same name.

MarcoRossignoli replied:
I discussed this approach with @jkotas to reuse existent `Privilege` objects https://github.com/dotnet/corefx/issues/17164#issuecomment-439629537
Do you have better idea?

stephentoub replied:
Do we allow this exception to propagate out  into user code?

MarcoRossignoli replied:
Not in theory for this namespace. At the moment there is only one place where we use exception https://github.com/dotnet/corefx/pull/33645/files#diff-466c09e7d1aad9f0599e4a112618b7a4R248
Is your concern about future `PrivilegeNotHeldException` misuse?

stephentoub replied:
> Is your concern about future PrivilegeNotHeldException misuse?

The concern would be if we allowed an instance of this exception to escape this library and into calling code.  If two types are defined with the same namespace/name but in different assemblies, they're still fundamentally two completely different types.  So if we're talking about an exception, and an instance of one is thrown and the other is specified in a catch block, the instance will pass right through the catch block because, although it happens to be named the same thing, it's a completely different type.  So, if this code threw a PrivilegeNotHeldException (this library's internal exception type of that name), and calling code had a catch block for PrivilegeNotHeldException (the public type exposed from another library), that catch block would not catch this exception, and in fact other than `catch (Exception)` to broadly catch all exception types, there'd be no specific way for the calling code to catch just this exception type, because it's internal.

That said, that concern isn't an issue if the exception never escapes this library, and that appears to be the case, so my concern doesn't apply.

MarcoRossignoli replied:
> That said, that concern isn't an issue if the exception never escapes this library, and that appears to be the case, so my concern doesn't apply.

Confirm "today", I'm aware that if someone in future "throw" this exception inside this project will be a problem.

stephentoub commented on src/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs:
Can we cache this array as a static readonly byte[]?

MarcoRossignoli replied:
I'll do it

stephentoub commented on src/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs:
```diff
@@ -208,20 +217,74 @@ public WindowsIdentity(string sUserPrincipalName)
+                    privilege = new Privilege("SeTcbPrivilege");
+                    privilege.Enable();
+                }
+                catch (PrivilegeNotHeldException) { }
```
If Enable throws, we'll still end up calling Revert later on.  Is that ok?  Presumably "yes", because that's what netfx does.

MarcoRossignoli replied:
https://github.com/dotnet/corefx/blob/a10890f4ffe0fadf090c922578ba0e606ebdd16c/src/System.Security.AccessControl/src/System/Security/AccessControl/Privilege.cs#L518 
https://github.com/dotnet/corefx/blob/a10890f4ffe0fadf090c922578ba0e606ebdd16c/src/System.Security.AccessControl/src/System/Security/AccessControl/Privilege.cs#L561

Revert will be skipped if state doesn't change.

stephentoub commented on src/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs:
```diff
@@ -1277,19 +1340,78 @@ private void AddTokenClaims(List<Claim> instanceClaims, TokenInformationClass to
+                    context.Undo();
+                    throw new SecurityException(SR.Argument_ImpersonateUser);
+                }
+            }
```
We're missing the code here that updates ExecutionContext.  That's important.  Is that coming in a subsequent change?  Developers expect to be able to write code like:
```C#
async Task DoWorkAsync(...)
{
    using (var impersonation = windowsIdentity.Impersonate())
    {
        ... // impersonating here
        await ...;
        ... // still impersonating here
    }
}
```
and have the impersonation flow across the await... the ExecutionContext is what makes that possible.

MarcoRossignoli replied:
>  Is that coming in a subsequent change? 

No I'm aware of it and asked how I can capture and flow without "delegate"(like `RunImpersonate`).
Can you follow my comments above https://github.com/dotnet/corefx/pull/33645#issuecomment-442385027 and enlighten me? Thank's @stephentoub!

stephentoub replied:
Why can't we just store the relevant information into an async local, and use the change callback to revert/restore the impersonation, as we do for run impersonated?
https://source.dot.net/#System.Security.Principal.Windows/System/Security/Principal/WindowsIdentity.cs,713

MarcoRossignoli replied:
Ok I'll work on it, as said above  there'll be some difference between `RunImpersonate/Impersonate` for netcoreapp and netfx...because netfx `RunImpersonate` is a wrapper on `WindowsImpersonationContext` that doesn't capture/revert context(an I don't understand how async/await works on netfx without call to [`ImpersonateLoggedOnUser`](https://referencesource.microsoft.com/#mscorlib/system/security/principal/windowsidentity.cs,756) is we come back on other thread), on current implementation of `RunImpersonate` EC is captured/reverted thank's to `Run`. But there isn't api to revert "current context" that I could capture and use on Undo(), what do you think,  am I wrong?

stephentoub replied:
> But there isn't api to revert "current context" that I could capture and use on Undo(), what do you think, am I wrong?

Can you just grab the current value when you (before you) impersonate and then store that value back in Undo?

MarcoRossignoli replied:
Ok for impersonation token value is not a problem, but for `CultureInfo/SynchronizationContext`....on `RunImpersonate` if I change `CurrentCulture` will be reverted on exit due to `Capture/Run`, to be consistent I should Capture/Restore(this is missing from EC api)

MarcoRossignoli commented:
Open points:
1) I added some tests using custom TaskScheduler to be sure that after "context flow" threads are impersonated in a correct way, because if we use ThreadPool there is a chance that we come back on already impersonated thread, hovewer I'm not sure that this makes sense...when we return an impersonated thread to ThreadPool do we cleanup token?On managed codebase I didn't found anything and I'm not so fluent on unmanaged one.
2) There is a difference between netfx/netcoreapp behaviour for `RunImpersonate` netcoreapp capture and revert context, netfx doesn't do it. Again there is a difference inside netcoreapp self between `RunImpersonate/WindowsImpersonationContext`, second ones doesn't capture/restore context on Undo(i.e. CurrentCulture will be not preserved) because we don't have `ExecutionContext.Restore(ExecutionContext ec)`(internal in netfx)
3) I copied `RunImpersonate` way to flow context with AsyncLocal, however I removed `Dispose` token on Undo because AsyncLocal callback are async and we need a valid handle to call [`ImpersonateLoggedOnUser`](https://github.com/dotnet/corefx/blob/master/src/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs#L760), is it ok?I mean nobody close this handles because are used in context flow chain, do we wait finalizer?

/cc @stephentoub  @jkotas @GrabYourPitchforks @danmosemsft

MarcoRossignoli commented:
@dotnet-bot test this please

MarcoRossignoli commented:
@dotnet-bot test this please

MarcoRossignoli commented:
@dotnet-bot test this please

MarcoRossignoli commented:
ping @stephentoub @jkotas @GrabYourPitchforks @danmosemsft

danmosemsft commented:
@stephentoub could you please take another look? I would like to get this merged before the holidays if we can.

MarcoRossignoli commented:
> @stephentoub could you please take another look? I would like to get this merged before the holidays if we can.

Here open points for clarity https://github.com/dotnet/corefx/pull/33645#issuecomment-443439465

danmosemsft commented:
`D:\j\workspace\windows-TGrou---33cbf18b\src\Common\src\Interop\Windows\sspicli\Interop.LsaRegisterLogonProcess.cs(14,9): error BCL0015: sspicli.dll!LsaRegisterLogonProcess is not supported on one\more targeted platforms. [D:\j\workspace\windows-TGrou---33cbf18b\src\System.Security.Principal.Windows\src\System.Security.Principal.Windows.csproj]` is because this APi is not in the UWP approved list. That's an omission I'll follow 

The other ones were added in https://github.com/dotnet/buildtools/pull/1615

We just need an exclusion file meantime.

MarcoRossignoli commented:
@danmosemsft `advapi32.dll!LsaNtStatusToWinError is not supported on one\more targeted platforms` same issue?Could I add it to same file?

MarcoRossignoli commented:
> Although, I would expect it to be already applied because of this file:

I did notice...

MarcoRossignoli commented on src/System.Security.Principal.Windows/tests/WindowsIdentityTests.Impersonate.cs:
```diff
@@ -0,0 +1,706 @@
+    [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsNotWindowsNanoServer))]
+    // On full framework 'RunImpersonate' doesn't capture/revert Execution context
+    [SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework)]
+    public void Impersonate_ExcutionContext_NotReverted()
```
done thank's

MarcoRossignoli commented:
>  In WindowsIdentityImpersonateFixture you should check for PlatformDetection.IsUap before trying to create or dispose users.

Removed tests on UAP because we need to create new users to properly run

karelz commented:
@danmosemsft all tests pass now. Is it ready for merge?

danmosemsft commented:
@karelz this is not an area I trust only my signoff. I suggest to wait for @stephentoub to return.

danmosemsft commented:
But, yay for getting green CI 🥂

danmosemsft commented:
@stephentoub do you have cycles to re-review?

stephentoub commented on src/System.Security.Principal.Windows/src/System/Security/Principal/WindowsImpersonationContext.cs:
```diff
@@ -0,0 +1,74 @@
+    public class WindowsImpersonationContext : IDisposable
+    {
+        private readonly SafeAccessTokenHandle _previousUserToken = SafeAccessTokenHandle.InvalidHandle;
+        private readonly AsyncLocal<SafeAccessTokenHandle> _currentImpersonatedToken;
```
Why is this needed? i.e. why can't this code just refer to s_currentImpersonatedToken (made internal instead of private)?

MarcoRossignoli replied:
can be done ok!

MarcoRossignoli replied:
done

stephentoub commented on src/System.Security.Principal.Windows/src/System/Security/Principal/WindowsImpersonationContext.cs:
```diff
@@ -0,0 +1,74 @@
+
+            if (isImpersonating)
+            {
+                // if we'are in a impersonated context we need to duplicate current token to allow revert to correct identity 
```
Nit: we'are => we're

stephentoub commented on src/System.Security.Principal.Windows/tests/WindowsIdentityTests.Impersonate.cs:
```diff
@@ -0,0 +1,709 @@
+    }
+
+    // Test to do in a domain without SeTcbPrivilege privilege
+    // [Fact]
```
What's the purpose / plan for these commented out [Fact]s?  Right now they're dead code.

MarcoRossignoli replied:
Because this tests are possible only inside a domain, I manually tested inside my office domain, but I've taked with @danmosemsft and he said that maybe some of you would re-test inside other domain, CI cannot test this cases. I mean this is manual tests.

bartonjs replied:
Crypto's solution to this: https://github.com/dotnet/corefx/blob/77096755443730962810662501f82550fdb4656d/src/System.Security.Cryptography.X509Certificates/tests/TestEnvironmentConfiguration.cs#L11-L12

Along with marking the tests as `[ConditionalFact(nameof(RunManualTests))]

bartonjs replied:
Of course, these tests require a bit more than uncommenting in order to work.

stephentoub commented on src/System.Security.Principal.Windows/tests/WindowsIdentityTests.Impersonate.cs:
```diff
@@ -0,0 +1,709 @@
+        List<bool> checkResult = new List<bool>();
+        string current = null;
+
+        await customFactory.StartNew(async () =>
```
This isn't behaving how you're hoping it will.  You need to add `.Unwrap()` at the end of the `StartNew(…)`.  The reason for that is StartNew doesn't have a `Func<Task>`-returning overload, only a `Func<T>`, so this is going to result in StartNew returning a `Task<Task>` rather than a `Task`, meaning you're not actually `await`'ing the inner Task but rather the outer one, such that this whole operation is going to complete the moment the async lambda first yields.  By adding `.Unwrap()`, you'll take that `Task<Task>` and "unwrap" it, returning a `Task` that represents the inner one, which will then be awaited.  See https://blogs.msdn.microsoft.com/pfxteam/2012/02/08/potential-pitfalls-to-avoid-when-passing-around-async-lambdas/.

MarcoRossignoli replied:
right thank's!

MarcoRossignoli replied:
I read that blog in past...but my memory made garbage :smile:

stephentoub commented on src/System.Security.Principal.Windows/tests/WindowsIdentityTests.Impersonate.cs:
```diff
@@ -0,0 +1,709 @@
+        List<bool> checkResult = new List<bool>();
+        string current = null;
+
+        await customFactory.StartNew(async () =>
```
Same for all of these.

MarcoRossignoli replied:
done

stephentoub commented on src/System.Security.Principal.Windows/tests/WindowsIdentityTests.Impersonate.cs:
```diff
@@ -0,0 +1,709 @@
+            ThreadsData.Add((InteropHelper.GetCurrentUserName(), Thread.CurrentThread.ManagedThreadId));
+
+            Interlocked.Decrement(ref _outstandingThread);
+        });
```
These should have { IsBackground = true }.

MarcoRossignoli replied:
ok

MarcoRossignoli replied:
done

stephentoub commented on src/System.Security.Principal.Windows/tests/WindowsIdentityTests.Impersonate.cs:
```diff
@@ -0,0 +1,709 @@
+// We skip all tests on Uap because we need to create some new Windows users before run
+// On Uap sandbox we cannot create users
+[SkipOnTargetFramework(TargetFrameworkMonikers.Uap)]
+public class WindowsIdentityTestsImpersonate : IClassFixture<WindowsIdentityImpersonateFixture>
```
These tests are modifying the current machine by adding user accounts.  They should all be outerloop.

MarcoRossignoli replied:
ok

MarcoRossignoli replied:
done

ViktorHofer replied:
/p:Outerloop=true

ViktorHofer replied:
https://github.com/dotnet/arcade/blob/master/src/Microsoft.DotNet.CoreFxTesting/build/core/test/Test.targets#L31

MarcoRossignoli replied:
strange command line it's ok(no -notrait category=OuterLoop)...but filter doesn't work
```
C:\git\corefx\src\System.Security.Principal.Windows (impersonate2 -> origin)                                                                                                                                                             
λ msbuild /t:rebuildandtest /v:m  /p:Outerloop=true                                                                                                                                                                                      
Microsoft (R) Build Engine version 15.9.21+g9802d43bc3 for .NET Framework                                                                                                                                                                
Copyright (C) Microsoft Corporation. All rights reserved.                                                                                                                                                                                
                                                                                                                                                                                                                                         
  System.Security.Principal.Windows -> C:\git\corefx\artifacts\bin\ref\System.Security.Principal.Windows\netstandard-Debug\System.Security.Principal.Windows.dll                                                                         
  System.Security.Principal.Windows -> C:\git\corefx\artifacts\bin\System.Security.Principal.Windows\netcoreapp-Windows_NT-Debug\System.Security.Principal.Windows.dll                                                                   
  System.Security.Principal.Windows.Tests -> C:\git\corefx\artifacts\bin\System.Security.Principal.Windows.Tests\netstandard-Windows_NT-Debug\System.Security.Principal.Windows.Tests.dll                                                
  Using C:\git\corefx\artifacts\bin\testhost\netcoreapp-Windows_NT-Debug-x64\ as the test runtime folder.                                                                                                                                
  Executing in C:\git\corefx\artifacts\bin\tests\System.Security.Principal.Windows.Tests\netcoreapp-Windows_NT-Debug-x64\                                                                                                                
  ----- start 17:31:35,20 ===============  To repro directly: =====================================================                                                                                                                      
  pushd C:\git\corefx\artifacts\bin\tests\System.Security.Principal.Windows.Tests\netcoreapp-Windows_NT-Debug-x64\                                                                                                                       
  set XUNIT_PERFORMANCE_MIN_ITERATION=1                                                                                                                                                                                                  
  set XUNIT_PERFORMANCE_MAX_ITERATION=1                                                                                                                                                                                                  
  set XUNIT_PERFORMANCE_MAX_ITERATION_INNER_SPECIFIED=1                                                                                                                                                                                  
  call C:\git\corefx\artifacts\bin\testhost\netcoreapp-Windows_NT-Debug-x64\\dotnet.exe xunit.console.dll System.Security.Principal.Windows.Tests.dll  -xml testResults.xml -notrait category=nonnetcoreapptests -notrait category=nonwin
  tests  -notrait category=failing                                                                                                                                                                                                       
  popd                                                                                                                                                                                                                                   
  ===========================================================================================================                                                                                                                            
  xUnit.net Console Runner v2.4.1-pre.build.4059 (64-bit .NET Core 4.6.27130.03)                                                                                                                                                         
    Discovering: System.Security.Principal.Windows.Tests (method display = ClassAndMethod, method display options = None)                                                                                                                
    Discovered:  System.Security.Principal.Windows.Tests (found 16 test cases)                                                                                                                                                           
    Starting:    System.Security.Principal.Windows.Tests (parallel test collections = on, max threads = 8)                                                                                                                               
    Finished:    System.Security.Principal.Windows.Tests                                                                                                                                                                                 
  === TEST EXECUTION SUMMARY ===                                                                                                                                                                                                         
     System.Security.Principal.Windows.Tests  Total: 18, Errors: 0, Failed: 0, Skipped: 0, Time: 0,212s                                                                                                                                  
  ----- end 17:31:37,12 ----- exit code 0 ----------------------------------------------------------                                                                                                                                     
                                                                                                                                                                                                                                         
C:\git\corefx\src\System.Security.Principal.Windows (impersonate2 -> origin)                                                                                                                                                             
λ msbuild /t:rebuildandtest /v:m                                                                                                                                                                                                         
Microsoft (R) Build Engine version 15.9.21+g9802d43bc3 for .NET Framework                                                                                                                                                                
Copyright (C) Microsoft Corporation. All rights reserved.                                                                                                                                                                                
                                                                                                                                                                                                                                         
  System.Security.Principal.Windows -> C:\git\corefx\artifacts\bin\ref\System.Security.Principal.Windows\netstandard-Debug\System.Security.Principal.Windows.dll                                                                         
  System.Security.Principal.Windows -> C:\git\corefx\artifacts\bin\System.Security.Principal.Windows\netcoreapp-Windows_NT-Debug\System.Security.Principal.Windows.dll                                                                   
  System.Security.Principal.Windows.Tests -> C:\git\corefx\artifacts\bin\System.Security.Principal.Windows.Tests\netstandard-Windows_NT-Debug\System.Security.Principal.Windows.Tests.dll                                                
  Using C:\git\corefx\artifacts\bin\testhost\netcoreapp-Windows_NT-Debug-x64\ as the test runtime folder.                                                                                                                                
  Executing in C:\git\corefx\artifacts\bin\tests\System.Security.Principal.Windows.Tests\netcoreapp-Windows_NT-Debug-x64\                                                                                                                
  ----- start 17:31:48,11 ===============  To repro directly: =====================================================                                                                                                                      
  pushd C:\git\corefx\artifacts\bin\tests\System.Security.Principal.Windows.Tests\netcoreapp-Windows_NT-Debug-x64\                                                                                                                       
  call C:\git\corefx\artifacts\bin\testhost\netcoreapp-Windows_NT-Debug-x64\\dotnet.exe xunit.console.dll System.Security.Principal.Windows.Tests.dll  -xml testResults.xml -notrait category=nonnetcoreapptests -notrait category=nonwin
  tests  -notrait category=OuterLoop -notrait category=failing                                                                                                                                                                           
  popd                                                                                                                                                                                                                                   
  ===========================================================================================================                                                                                                                            
  xUnit.net Console Runner v2.4.1-pre.build.4059 (64-bit .NET Core 4.6.27130.03)                                                                                                                                                         
    Discovering: System.Security.Principal.Windows.Tests (method display = ClassAndMethod, method display options = None)                                                                                                                
    Discovered:  System.Security.Principal.Windows.Tests (found 16 test cases)                                                                                                                                                           
    Starting:    System.Security.Principal.Windows.Tests (parallel test collections = on, max threads = 8)                                                                                                                               
    Finished:    System.Security.Principal.Windows.Tests                                                                                                                                                                                 
  === TEST EXECUTION SUMMARY ===                                                                                                                                                                                                         
     System.Security.Principal.Windows.Tests  Total: 18, Errors: 0, Failed: 0, Skipped: 0, Time: 0,207s                                                                                                                                  
  ----- end 17:31:49,95 ----- exit code 0 ----------------------------------------------------------                                                                                                                                     
                                                                                                                                                                                                                                         
C:\git\corefx\src\System.Security.Principal.Windows (impersonate2 -> origin)                                                                                                                                                             
λ                                                                                                                                                                                                                                        
```
Surely I missing something...however I use CI don't worry thank's for help!

ViktorHofer replied:
which filter do you mean? looking if we have a bug somewhere :)

ViktorHofer replied:
Also where is this coming from?

```
set XUNIT_PERFORMANCE_MIN_ITERATION=1                                                                                                                                                                                                  
  set XUNIT_PERFORMANCE_MAX_ITERATION=1                                                                                                                                                                                                  
  set XUNIT_PERFORMANCE_MAX_ITERATION_INNER_SPECIFIED=1         
```

I removed that months ago, I'm surprised these commands are in your logs.

MarcoRossignoli replied:
This branch is very old...maybe this is the issue. However I wanted run only Outerloop tests, I'll wait ongoing discussion and in case I'll rebase this branch to try.

stephentoub commented:
@bartonjs, can you review this, too, please?  Thanks.

MarcoRossignoli commented:
@stephentoub @bartonjs I would point out my concerns before merge https://github.com/dotnet/corefx/pull/33645#issuecomment-443439465
If you can take a look and answer I'll be more comfortable

MarcoRossignoli commented:
> I'm not entirely following. How would ExecutionContext.Restore help?

In case of `RunImpersonate` we run a delegate so we can capture context(list of AsyncLocal) and on `Run` exits we can "restore" context(with AsyncLocal events). But if we use simple WindowsImpersonationContext we cannot "capture" context before "Impersonate" and restore after...because we "run" a body and not a delegate...how we can "restore" captured context?i.e.
```cs
WindowsIdentity.RunImpersonated(token, () => // <- we capture ExecutionContext thank's to ExecutionContext.Run(capturedContext, delegate)
{
     // for instance change current culture
}
// here we are sure that context is restored thank's to ExecutionContext.Run(...) restore and we have old culture before impersonate
```
but
```cs
using (WindowsImpersonationContext ctxUser1 = CtxFactoryUser1()) // <- we could call ExecutionContext.Capture()
{
    // for instance change current culture
}
// here how we can restore old context and reset culture on execution context disposing?
```

> Now that we have Impersonate and WindowsImpersonationContext, why can't RunImpersonated just do effectively the same thing that netfx does:

It was my first idea...we could do that if you agree, we talked about it in past with @jkotas https://github.com/dotnet/corefx/issues/17164#issuecomment-440014067 https://github.com/dotnet/corefx/issues/17164#issuecomment-440342585
(however we are changing current execution context behaviour, right?)

> We do clean it up. All...

Understood we should be safe thank's to `s_currentImpersonatedToken` that in case of async/await will revert token on exit `Interop.Advapi32.RevertToSelf()`. My idea of custom thread pool is to be sure that I cannot impersonate a async/await thread pool thread(windows token) and after for other async/await we get same impersonated thread and generate false positive, but you're right better use `TaskCreationOptions.LongRunning along with TaskScheduler.Default`

MarcoRossignoli commented:
@dotnet-bot test Outerloop Windows x86 Release Build please
@dotnet-bot test Outerloop NETFX x86 Debug Build
@dotnet-bot test Outerloop Windows x64 Debug Build

bartonjs commented on src/System.Security.AccessControl/tests/System.Security.AccessControl.Tests.csproj:
```diff
@@ -1,8 +1,9 @@
     <ProjectGuid>{70FAC855-CAC6-4523-8477-880548D58A1B}</ProjectGuid>
     <UnsupportedPlatforms>Linux;NetBSD;OSX</UnsupportedPlatforms>
     <Configurations>netstandard-Windows_NT-Debug;netstandard-Windows_NT-Release</Configurations>
+    <DefineConstants>$(DefineConstants);ACCESSCONTROL_ASSEMBLY</DefineConstants>
```
Why would the tests declare this?  They don't seem to be including the shared source that uses this value.

bartonjs commented on src/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs:
```diff
@@ -208,20 +217,72 @@ public WindowsIdentity(string sUserPrincipalName)
+                    using (SafeLocalAllocHandle sourceNameBuffer = Interop.Kernel32.LocalAlloc(0, new UIntPtr(sourceNameLength)))
+                    {
+                        if (sourceNameBuffer.IsInvalid)
+                            throw new OutOfMemoryException();
```
Missing braces (this if is contained within a braced block, therefore braces are required).

bartonjs commented on src/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs:
```diff
@@ -208,20 +217,72 @@ public WindowsIdentity(string sUserPrincipalName)
+        /// </summary>
+        /// <param name="logonProcessNameBuffer"></param>
+        /// <returns></returns>
+        private static SafeLsaHandle GetLsaHandle(byte[] logonProcessNameBuffer)
```
If there's only one caller to this method is it worth taking the parameter, or should it just be a static field reference within the method?

bartonjs commented on src/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs:
```diff
@@ -710,7 +771,7 @@ public void Dispose()
         //
 
-        private static AsyncLocal<SafeAccessTokenHandle> s_currentImpersonatedToken = new AsyncLocal<SafeAccessTokenHandle>(CurrentImpersonatedTokenChanged);
+        internal static AsyncLocal<SafeAccessTokenHandle> s_currentImpersonatedToken = new AsyncLocal<SafeAccessTokenHandle>(CurrentImpersonatedTokenChanged);
```
I don't see any reassignment of this field; can it be made `readonly`?  If not, I'd rather make an internal accessor for the tests than give them write access to a field.

bartonjs replied:
(tests, peer type, whatever; it still seems like the `readonly` modifier can be applied, since `AsyncLocal` is a reference type that never gets reassigned, only mutated)

bartonjs commented on src/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs:
This is the only call to `LookupAuthenticationPackage`, which means that `Interop.SspiCli.AuthenticationPackageNames.MICROSOFT_KERBEROS_NAME_A` is the only argument ever provided, which means that every call to Ascii.GetBytes will produce the same value (in a different array).

Maybe this value should be pre-ASCII'd and held for the lifetime of the process?

bartonjs commented on src/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs:
```diff
@@ -208,20 +217,72 @@ public WindowsIdentity(string sUserPrincipalName)
 
```
Just above this are two places that can throw; theoretically they should also dispose accessTokenHandle.

bartonjs commented on src/System.Security.Principal.Windows/src/System/Security/Principal/WindowsImpersonationContext.cs:
```diff
@@ -0,0 +1,70 @@
+                }
+
+                // reset AsyncLocal to allow impersonation context flow in case of async/await
+                WindowsIdentity.s_currentImpersonatedToken.Value = _previousUserToken;
```
This seems like it should be a method (which will probably get inlined) rather than direct field manipulation.

bartonjs commented on src/System.Security.Principal.Windows/src/System/Security/Principal/WindowsImpersonationContext.cs:
```diff
@@ -0,0 +1,70 @@
+
+            if (isImpersonating)
+            {
+                // if we're in a impersonated context we need to duplicate current token to allow revert to correct identity 
```
The _previousUserToken field is always initialized to a new object (`SafeAccessTokenHandle.InvalidHandle` returns new every time).  Either dispose it here before reassignment, or restructure the field to be set in an else here.

bartonjs commented on src/System.Security.Principal.Windows/src/System/Security/Principal/WindowsImpersonationContext.cs:
```diff
@@ -0,0 +1,70 @@
+                if (_previousUserToken != null && !_previousUserToken.IsClosed)
+                {
+                    Undo();
+                }
```
Seems to be missing the `_previousUserToken.Dispose()` that was present in netfx.

bartonjs commented on src/System.Security.Principal.Windows/tests/WindowsIdentityTests.Impersonate.cs:
```diff
@@ -0,0 +1,654 @@
+// We skip all tests on Uap because we need to create some new Windows users before run
+// On Uap sandbox we cannot create users
+[SkipOnTargetFramework(TargetFrameworkMonikers.Uap)]
+public class WindowsIdentityTestsImpersonate : IClassFixture<WindowsIdentityImpersonateFixture>
```
Is there a reason this class isn't in a namespace?

blowdart commented:
Wait, I thought in previous discussions we decided not to support this, and only support the safer, new syntax which forces the use of using? Why is this up for discussion again as a PR, rather than a wider "Do we want to bring back an API that people used disastrously wrong?"

bartonjs commented on src/System.Security.Principal.Windows/tests/WindowsIdentityTests.Impersonate.cs:
```diff
@@ -0,0 +1,654 @@
+        }
+    }
+
+    [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsNotWindowsNanoServer))]
```
This should already be covered by the attribute on the class, no?

bartonjs commented on src/System.Security.Principal.Windows/tests/WindowsIdentityTests.Impersonate.cs:
```diff
@@ -0,0 +1,654 @@
+        string previous = InteropHelper.GetCurrentUserName();
+
+        // Assert.NotEqual() lacks ignoreCase overload
+        Assert.False(previous.Equals(impersonatedAccount.AccountName, StringComparison.InvariantCultureIgnoreCase));
```
`Assert.Equal(previous, impersonatedAccount.Name, StringComparer.InvariantCultureIgnoreCase);`

bartonjs commented on src/System.Security.Principal.Windows/tests/WindowsIdentityTests.Impersonate.cs:
```diff
@@ -0,0 +1,654 @@
+        string previous = InteropHelper.GetCurrentUserName();
+
+        // impersonating a zero token means clear the token on the thread in this case NOP
+        using (WindowsIdentity.Impersonate(SafeAccessTokenHandle.InvalidHandle.DangerousGetHandle()))
```
Why not just say `IntPtr.Zero`?

bartonjs commented:
> Wait, I thought in previous discussions we decided not to support this, and only support the safer, new syntax which forces the use of using? 

If that was a conscious choice, and the only goal of the issue is to "make it so no MissingMethodException is raised", then all of this could easily just be a PlatformNotSupportedException.

FWIW, I'm not a fan of impersonation APIs, so I'd support the PNSE.

blowdart commented:
It was a conscious choice in one of the 3.0 compat meetings. @terrajobst did you have this documented somewhere?

stephentoub commented:
> Why is this up for discussion again as a PR, rather than a wider

https://github.com/dotnet/corefx/issues/17164

blowdart commented:
Aha

blowdart commented:
Closing as this API is not returning.
